proposals/NNNN-refining-property-wrapper-related-initialization.md

```diff
@@ -0,0 +1,263 @@
+
+## Effect on ABI Stability
+
+This feature and the features it depends on are either entirely non-public or implemented entirely in terms of frontend transformations and do not have an impact on ABI.
+
+## Effect on API Resilience
+
+Because the synthesized memberwise initializer is always non-`public`, this proposal does not introduce any functionality that would affect API resilience.
+
+## Alternatives Considered
+
+*TBC*
+
+## Future Directions
+
+### Inline projected value initialization
+
+The `$`-initialization syntax could be extended to allow for its use in the declaration of the wrapped property itself, e.g.:
+
+```swift
+struct S {
+  @Wrapper
+  var $property: Wrapper = someProjectedValue
+}
+```
+
+would expand to:
+
+```swift
+struct S {
+  private var _property: Wrapper = .init(projectedValue: someProjectedValue)
+    
+  var property: Int {
+    get { _property.wrappedValue }
+    set { _property.wrappedValue = newValue }
+  }
+}
+```
+
+The authors elect to exclude such a construction from this proposal. The suggested syntax above tries to do many things at once:
+1. Declare a property named `property`.
+2. Declare a wrapper/storage type `Wrapper` for that property.
+3. Declare the type of the projected value.
+4. Declare that the property should be initialized via `init(projectedValue:)`.
+5. Declare the projected value to be passed to `init(projectedValue:)`.
+
+Notably, this syntax does not provide an easy place to indicate the type of `property` itself. The reader may be able to infer it based on the type of the wrapper or the projected value, but this could require inspection of the type declaration.
+
+
+While it is possible that the syntax here could be massaged into something that made a bit more sense, there are enough open questions that the authors would rather see this form receive further, separate consideration.
+
+### Accept the Backing Storage in Parameters
+
+[SE 0293](https://github.com/apple/swift-evolution/blob/main/proposals/0293-extend-property-wrappers-to-function-and-closure-parameters.md) settled on parameters accepting either the wrapped or projected value. If neither was supported, the resulting declaration would be uncallable. That can be limiting, though, especially for memberwise initializers:
+
+```swift
+@propertyWrapper struct BackingStorageOnly {
+  let wrappedValue = 0
+}
+
+struct Client {
+  // ℹ️ `@BackingStorageOnly` doesn't declare an `init(wrappedValue:)` or `init(projectedValue:)` initializer.
+  @BackingStorageOnly var property: Int 
+  
+  // ❌ Invoking uncallable memberwise initializer.
+  static let `default` = Client()
+}
+```
+
+The rationale was that the _private_ backing storage must not be exposed to function clients. This is a valid concern, but doesn't preclude a `private`, backing-storage-accepting function that follows these rules:
+
+* A wrapper with neither an `init(wrappedValue:)` or `init(projectedValue:)` special initializer is considered **private API level**;
```

Comment:
Although not in our original discussion, I think _private_-API wrappers would fit in quite well with the SE-293-defined API significance characterizations (API level, implementation detail). Thoughts?

Reply:
Do we run afoul of this "A property wrapper will only be inferred as API if init(projectedValue:) is declared directly in the nominal property wrapper type." from SE-0293?

Reply:
I think we could definitely come up with rules for private-API-wrappers here that make sense and are compatible with the old API-wrapper rule, but since this is being left as a future direction right now probably not worth it to worry about it too much. IMO what you have here is good grounding for future discussion of this feature.

Reply:
That's a good point, but I agree we don't have to think of all the semantics. Nonetheless, we can list it as an open question:

> One issue with introducing a new API significance characterizations that combines API and implementation detail is how it will be inferred. This is not a straightforward decision, because API-level inference depends solely on the primary wrapper type declaration; `init(wrappedValue:)` functions as a special wrapper init _even_ in extensions.